test/integration/validation/chain_validator_storage_test.cpp

```diff
@@ -0,0 +1,162 @@
+              .signAndAddSignature(keys.at(0)));
+
+      auto ms = createMutableStorage();
+
+      ms->apply(block);
+      storage->commit(std::move(ms));
+
+      return block;
+    }
+
+    /// Create an observable from chain and return its validation status
+    auto createAndValidateChain(
+        std::vector<std::shared_ptr<shared_model::interface::Block>> chain) {
+      auto ms = createMutableStorage();
+      return validator->validateChain(rxcpp::observable<>::iterate(chain), *ms);
+    }
+
+    std::shared_ptr<validation::ChainValidatorImpl> validator;
+    std::vector<shared_model::crypto::Keypair> keys;
+  };
+
+  /**
+   * @given initialized storage with 4 peers, second block with add peer
+   * command, third block signed by new peer, blocks are signed by supermajority
+   * of ledger peers
+   * @when chain with second and third blocks is validated
+   * @then result is successful
+   */
+  TEST_F(ChainValidatorStorageTest, PeerAdded) {
+    auto block1 = generateAndApplyFirstBlock();
+
+    auto add_peer =
+        completeTx(baseTx().addPeer("0.0.0.0:50545", keys.at(4).publicKey()));
+    auto block2 = completeBlock(baseBlock({add_peer}, 2, block1.hash())
+                                    .signAndAddSignature(keys.at(0))
+                                    .signAndAddSignature(keys.at(1))
+                                    .signAndAddSignature(keys.at(2)));
+
+    auto block3 = completeBlock(baseBlock({dummyTx(3)}, 3, block2.hash())
+                                    .signAndAddSignature(keys.at(0))
+                                    .signAndAddSignature(keys.at(1))
+                                    .signAndAddSignature(keys.at(2))
+                                    .signAndAddSignature(keys.at(3))
+                                    .signAndAddSignature(keys.at(4)));
+
+    ASSERT_TRUE(createAndValidateChain({clone(block2), clone(block3)}));
+  }
+
+  /**
+   * @given initialized storage with 4 peers, second and third blocks signed by
+   * supermajority of ledger peers
+   * @when chain with second and third blocks is validated
+   * @then result is successful
+   */
+  TEST_F(ChainValidatorStorageTest, NoPeerAdded) {
+    auto block1 = generateAndApplyFirstBlock();
+
+    auto block2 = completeBlock(baseBlock({dummyTx(2)}, 2, block1.hash())
+                                    .signAndAddSignature(keys.at(0))
+                                    .signAndAddSignature(keys.at(1))
+                                    .signAndAddSignature(keys.at(2)));
+
+    auto block3 = completeBlock(baseBlock({dummyTx(3)}, 3, block2.hash())
+                                    .signAndAddSignature(keys.at(0))
+                                    .signAndAddSignature(keys.at(1))
+                                    .signAndAddSignature(keys.at(2))
+                                    .signAndAddSignature(keys.at(3)));
+
+    ASSERT_TRUE(createAndValidateChain({clone(block2), clone(block3)}));
+  }
+
+}  // namespace iroha
```

Comment:
Can you add some faulty case, when peer checks malicious chain with absent supermajority or with wrong hash?